[Bug] FPKGi feed: category-based entry names aren't unique within a ROM, so packages overwrite each other

**RomM version**

5.1.0 (Docker)

**Describe the bug**

Follow-up to #3962, fixed by #3966. Per-file entries work now — thanks. Testing on 5.1.0 surfaced two problems.

**1. Category-based entry names are not unique within a ROM (blocking)**

The naming rule from #3966 replaces the file name with the category label when a file has a scanned category, producing `<Game> - Update` or `<Game> - DLC`. When a ROM contains more than one package in the same category, every one of them gets an identical `name`. This is critical for DLC Files.

Two consequences:

- The entries are indistinguishable in FPKGi. Same name, same `title_id`, and often the same `size`, so there is no way to tell which package is which.
- FPKGi builds its destination path as `[<title_id>] <name>.pkg`. Identical names therefore resolve to the same file on disk, and each download silently overwrites the previous one.

Feed output (trimmed) — two distinct packages, identical `name` and `title_id`:

```json
"http://romm:8080/api/roms/137/files/content/The%20Last%20of%20Us%E2%84%A2%20Remastered%20-%20DLC-Early%20Brawler%20Survival%20Skill-CUSA00556.pkg": {
  "name": "The Last of Us Remastered - DLC",
  "size": 1048576,
  "title_id": "CUSA00556"
},
"http://romm:8080/api/roms/138/files/content/The%20Last%20of%20Us%E2%84%A2%20Remastered%20-%20DLC-Loadout%20Points-CUSA00556.pkg": {
  "name": "The Last of Us Remastered - DLC",
  "size": 1048576,
  "title_id": "CUSA00556"
}
```

FPKGi log — two different source URLs, one destination path:

```
11:11:19PM [DEBUG] Setting CURL options with URL: .../api/roms/137/files/content/The%20Last%20of%20Us%E2%84%A2%20Remastered%20-%20DLC-Early%20Brawler%20Survival%20Skill-CUSA00556.pkg
11:11:19PM [DEBUG] Download completed successfully: /user/data/pkg/fpkgi/[CUSA00556] The Last of Us Remastered - DLC.pkg

11:11:23PM [INFO]  Attempting to download PKG from: .../api/roms/138/files/content/The%20Last%20of%20Us%E2%84%A2%20Remastered%20-%20DLC-Loadout%20Points-CUSA00556.pkg
11:11:23PM [DEBUG] Downloading file and writing to: /user/data/pkg/fpkgi/[CUSA00556] The Last of Us Remastered - DLC.pkg
11:11:23PM [DEBUG] Download completed successfully: /user/data/pkg/fpkgi/[CUSA00556] The Last of Us Remastered - DLC.pkg
```

**2. Installing one package routes other packages of the same ROM to the install path**

With a base game and an update sharing `title_id: CUSA07208`:

- Download base → download update → install base → install update: works.
- Download base → install base → select update: fails.

In the failing case FPKGi skips the download entirely and goes straight to installing from disk, for a file that was never downloaded:

```
11:46:12PM [INFO]  Attempting to download PKG from: .../api/roms/123/files/content/Human.Fall.Flat.v1.01.PATCH.PS4-CUSA07208.pkg
11:46:12PM [INFO]  Installing package [/user/data/pkg/fpkgi/[CUSA07208] Human: Fall Flat - Update.pkg] from /user/data/pkg/fpkgi/ & keeping file.
11:46:12PM [DEBUG] Checking if file exists: /user/data/pkg/fpkgi/[CUSA07208] Human: Fall Flat - Update.pkg
11:46:12PM [ERROR] Failed to open file: /user/data/pkg/fpkgi/[CUSA07208] Human: Fall Flat - Update.pkg
11:46:12PM [ERROR] Package installation has failed.
```

For contrast, the base game on the same run took the download path normally:

```
11:43:22PM [INFO]  Background download has begun.
11:43:22PM [DEBUG] Downloading file and writing to: /user/data/pkg/fpkgi/[CUSA07208] Human: Fall Flat - Human.Fall.Flat.PS4-CUSA07208.pkg
11:43:22PM [DEBUG] No resume file found, creating new...
11:43:22PM [DEBUG] New file created successfully.
11:43:58PM [DEBUG] Download completed successfully: ...
```

The base and update entries differ in both `name` and download URL; the only field they share is `title_id`. I can't confirm how FPKGi decides between downloading and installing, so I'm reporting the observed routing rather than a cause — but the shared title ID is the only candidate visible in the feed. The Greptile review on #3966 raised a related concern about `title_id` uniqueness and installation status before merge. ( --> is it a FPKGi Issue?)

**To Reproduce**

Naming collision:

1. Create a PS4 ROM folder with a base game plus several DLC packages in a `dlc` subfolder (several packages in an `update` subfolder reproduces it equally). Scan it.
2. Request `/api/feeds/fpkgi/ps4` — all entries in the same category share one `name` and one `title_id`.
3. Point FPKGi's `CONTENT_URLS` at the feed and download one DLC entry, then a different DLC entry from the same game.
4. Only one `.pkg` exists in the download folder; the second overwrote the first.

Install routing:

1. Clear the download folder.
2. Download the base game, then install it.
3. Select the update → FPKGi attempts to install from disk instead of downloading, and fails.
4. Downloading both packages before installing either one avoids this.

Both behaviors are identical with a single `games` URL and with the four-URL `content_type` split.


**Server:**

- OS: Debian LXC
- Version: RomM 5.1.0, Docker

**Client:**

- Device: PS4 Pro (CFW + GoldHEN)
- OS: 11.00
- Version: FPKGi 1.10.0

**Additional context**

Affects the FPKGi feed endpoint only; the RomM web UI is unaffected. Both issues involve fields that are resolved once per ROM and reused for every package it contains.

--- a/backend/endpoints/feeds.py
+++ b/backend/endpoints/feeds.py
@@ -1,6 +1,7 @@
 import csv
 import io
 import re
+from collections import Counter
 from collections.abc import Sequence
 from datetime import datetime
 from typing import Annotated
@@ -582,14 +583,40 @@ FPKGI_CATEGORY_LABELS: dict[RomFileCategory, str] = {
 FPKGI_TITLE_ID_REGEX = re.compile(r"(?:CUSA|PPSA)\d{5}", re.IGNORECASE)
 
 
-def fpkgi_item_name(rom: Rom, file: RomFile, *, is_single_file: bool) -> str:
-    """Name shown in FPKGi, disambiguated when a rom holds several packages."""
-    rom_name = rom.name or rom.fs_name
-    if is_single_file:
-        return rom_name
-
+def fpkgi_name_candidates(rom_name: str, file: RomFile) -> list[str]:
+    """Names for a package, from the most readable to the most specific."""
     label = FPKGI_CATEGORY_LABELS.get(file.category) if file.category else None
-    return f"{rom_name} - {label or file.file_name_no_ext}"
+    stem = file.file_name_no_ext
+
+    candidates = []
+    if label:
+        candidates.append(f"{rom_name} - {label}")
+    candidates.append(f"{rom_name} - {stem}")
+    if label:
+        candidates.append(f"{rom_name} - {label} - {stem}")
+    # Last resort: two categories can hold packages with the same file name
+    candidates.append(f"{rom_name} - {stem} ({file.id})")
+    return candidates
+
+
+def fpkgi_item_names(rom: Rom, files: list[RomFile]) -> dict[int, str]:
+    """Name shown in FPKGi per package, keyed by rom file id.
+
+    FPKGi downloads to `[<title_id>] <name>.pkg`, and packages of a rom share a
+    title id, so two of them sharing a name overwrite each other on the console.
+    Each package therefore takes the first of its candidate names that no other
+    package in the rom lays claim to.
+    """
+    rom_name = rom.name or rom.fs_name
+    if len(files) == 1:
+        return {files[0].id: rom_name}
+
+    candidates = {f.id: fpkgi_name_candidates(rom_name, f) for f in files}
+    claimed = Counter(name for names in candidates.values() for name in names)
+    return {
+        file_id: next(name for name in names if claimed[name] == 1)
+        for file_id, names in candidates.items()
+    }
 
 
 def fpkgi_title_id(rom: Rom, files: list[RomFile]) -> str:
@@ -656,6 +683,9 @@ def fpkgi_feed(
             if f.file_extension.lower() == "pkg" and not f.missing_from_fs
         ]
         title_id = fpkgi_title_id(rom, pkg_files)
+        # Named over every package of the rom, not just the filtered ones, so a
+        # name means the same thing whichever CONTENT_URLS slot serves it
+        item_names = fpkgi_item_names(rom, pkg_files)
         cover_url = (
             str(URLPath(rom.path_cover_large).make_absolute_url(request.base_url))
             if rom.path_cover_large
@@ -670,7 +700,7 @@ def fpkgi_feed(
 
             download_url = generate_romfile_download_url(request, file)
             response_data[download_url] = FPKGiFeedItemSchema(
-                name=fpkgi_item_name(rom, file, is_single_file=len(pkg_files) == 1),
+                name=item_names[file.id],
                 size=file.file_size_bytes,
                 title_id=title_id,
                 region=rom.regions[0] if rom.regions else None,

--- a/backend/tests/endpoints/feeds.py
+++ b/backend/tests/endpoints/feeds.py
@@ -327,6 +327,79 @@ def test_fpkgi_feed_multi_file_rom(
     assert response.status_code == status.HTTP_400_BAD_REQUEST
 
 
+def test_fpkgi_feed_names_are_unique_within_a_rom(
+    client: TestClient, access_token: str, platform: Platform, rom: Rom
+):
+    platform = db_platform_handler.update_platform(
+        platform.id, {"name": "PlayStation 4", "slug": UPS.PS4, "fs_slug": UPS.PS4}
+    )
+    rom = db_rom_handler.update_rom(
+        rom.id,
+        {
+            "platform_id": platform.id,
+            "name": "Test PS4",
+            "fs_name": "Test PS4",
+            "fs_name_no_tags": "Test PS4",
+            "fs_name_no_ext": "Test PS4",
+            "fs_extension": "",
+            "fs_path": f"{platform.slug}/roms",
+            "fs_size_bytes": 369,
+            "regions": ["US"],
+        },
+    )
+    for sub_path, file_name, category in (
+        ("", "Test PS4 base.pkg", None),
+        ("update", "Test PS4 patch.pkg", RomFileCategory.UPDATE),
+        ("dlc", "Test PS4 brawler.pkg", RomFileCategory.DLC),
+        ("dlc", "Test PS4 loadout.pkg", RomFileCategory.DLC),
+        # Same file name in two categories, so the file name alone is ambiguous
+        ("dlc", "Test PS4 extra.pkg", RomFileCategory.DLC),
+        ("demo", "Test PS4 extra.pkg", RomFileCategory.DEMO),
+        ("demo", "Test PS4 trial.pkg", RomFileCategory.DEMO),
+    ):
+        db_rom_handler.add_rom_file(
+            RomFile(
+                rom_id=rom.id,
+                file_name=file_name,
+                file_path=f"{rom.fs_path}/{rom.fs_name}/{sub_path}".rstrip("/"),
+                file_size_bytes=123,
+                category=category,
+            )
+        )
+
+    response = client.get(
+        "/api/feeds/fpkgi/ps4",
+        headers={"Authorization": f"Bearer {access_token}"},
+    )
+    assert response.status_code == status.HTTP_200_OK
+
+    data = response.json()["DATA"]
+    assert len(data) == 7
+    assert sorted(entry["name"] for entry in data.values()) == [
+        "Test PS4 - DLC - Test PS4 extra",
+        "Test PS4 - Demo - Test PS4 extra",
+        "Test PS4 - Test PS4 base",
+        "Test PS4 - Test PS4 brawler",
+        "Test PS4 - Test PS4 loadout",
+        "Test PS4 - Test PS4 trial",
+        "Test PS4 - Update",
+    ]
+
+    # Filtering must not change the name a package is served under
+    response = client.get(
+        "/api/feeds/fpkgi/ps4?content_type=dlc",
+        headers={"Authorization": f"Bearer {access_token}"},
+    )
+    assert response.status_code == status.HTTP_200_OK
+
+    dlc_data = response.json()["DATA"]
+    assert sorted(entry["name"] for entry in dlc_data.values()) == [
+        "Test PS4 - DLC - Test PS4 extra",
+        "Test PS4 - Test PS4 brawler",
+        "Test PS4 - Test PS4 loadout",
+    ]
+
+
 def test_kekatsu_feed(
     client: TestClient, access_token: str, platform: Platform, rom: Rom
 ):

